### PR TITLE
Unet3+ added

### PR DESCRIPTION
**Pull request by ayu-22**
Hey @qubvel, I have implemented the Unet3+ architecture as asked in issue #458.

Please review my pull request

### CODE REVIEW

**Comment by qubvel** [comment]
Please rebase your commits to the current master branch, it is not possible to merge this request

**Comment by ayu-22** [comment]
@qubvel I have changed the base to master

File: segmentation_models_pytorch/unet3plus/model.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            encoder_channels=self.encoder.out_channels,
+            decoder_channels=(256, 128, 64, 32, 16),
+            n_blocks=encoder_depth,
+            use_batchnorm=decoder_use_batchnorm,
+            center=True if encoder_name.startswith("vgg") else False,
+            attention_type=decoder_attention_type,
+        )
+
+        self.segmentation_head = SegmentationHead(
+            in_channels=decoder_channels[encoder_depth-1]*encoder_depth,
```
Comment (qubvel) [inline segmentation_models_pytorch/unet3plus/model.py:81]: please add spaces
[encoder_depth - 1] * encoder_depth,

File: segmentation_models_pytorch/unet3plus/decoder.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        
+        x = self.center(head)  #first tensor is input into centr block
+
+        decoded_features = []
+        decoded_features.append(x)
+        
+        #tensors are calculated for each stage
+        for i, stage in enumerate(self.stages):
+      
+          total_features = current_features.copy()
```
Comment (qubvel) [inline segmentation_models_pytorch/unet3plus/decoder.py:237]: Why do we need a copy here?

File: segmentation_models_pytorch/unet3plus/decoder.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        # computing blocks input and output channels
+        head_channels = encoder_channels[0]
+        in_channels = encoder_channels[1:]
+        in_channels = in_channels[::-1] #reversing the list
+
+        out_channels = [head_channels] + list(decoder_channels) 
+        previous_channels = [i * n_blocks for i in out_channels] #calculating previos channels i.e concatenated output channels of previous block
+        previous_channels[0] = int(previous_channels[0]/n_blocks) #first tensor as it comes from CenterBlock, hence no concatenation
+
+        if True:
```
Comment (qubvel) [inline segmentation_models_pytorch/unet3plus/decoder.py:199]: Looks strange: `else` statement for `if True`

File: README.md [hunk 4]
```diff
@@ -77,6 +77,7 @@ Congratulations! You are done! Now you can train your model with your favorite f
 #### Architectures <a name="architectures"></a>
  - Unet [[paper](https://arxiv.org/abs/1505.04597)] [[docs](https://smp.readthedocs.io/en/latest/models.html#unet)]
  - Unet++ [[paper](https://arxiv.org/pdf/1807.10165.pdf)] [[docs](https://smp.readthedocs.io/en/latest/models.html#id2)]
+ - Unet3+ [[paper](https://arxiv.org/ftp/arxiv/papers/2004/2004.08790.pdf)]
```
Comment (qubvel) [inline README.md:80]: Add to docs page and link

**Review by qubvel — COMMENTED** [review]
Please add Unet3+ to the forward-backward tests

**Comment by qubvel** [comment]
Did you try to train this network? Any comparison to Unet?

File: segmentation_models_pytorch/unet3plus/decoder.py (hunk 2 above)
Comment (ayu-22) [reply segmentation_models_pytorch/unet3plus/decoder.py:237]: We need a copy here because we don't want to change the values of current_features when we are updating total_features.

File: segmentation_models_pytorch/unet3plus/model.py (hunk 1 above)
Comment (ayu-22) [reply segmentation_models_pytorch/unet3plus/model.py:81]: Okay I will update that

File: segmentation_models_pytorch/unet3plus/decoder.py (hunk 3 above)
Comment (ayu-22) [reply segmentation_models_pytorch/unet3plus/decoder.py:199]: Yeah, it should be center. I will update that

File: README.md (hunk 4 above)
Comment (ayu-22) [reply README.md:80]: Okay

**Comment by ayu-22** [comment]
> Did you try to train this network? Any comparison to Unet?

Currently, I am doing the comparison with respect to Unet. I will let you know, once I am done with experiments

File: segmentation_models_pytorch/unet3plus/decoder.py (hunk 2 above)
Comment (qubvel) [reply segmentation_models_pytorch/unet3plus/decoder.py:237]: Not sure if you make a copy of a `list` and tensors will be copied also

Comment (ayu-22) [reply segmentation_models_pytorch/unet3plus/decoder.py:237]: I have checked that, when I am making the copy of the list consisting of tensors, the tensors are also getting copied.
For e.g., if let say, LIS is a list consisting of 5 tensors of size (64,64). 
LIS_new = LIS.copy()
here, a new list LIS_new will be created which will consist of 5 tensors of size (64,64).

If, I just did LIS_new = LIS, then no new list was getting created, only a new reference for LIS was given. i.e changes done on LIS_new will get reflected in LIS.

But, that is not the case when I do LIS_new = LIS

That is why I have used copy()

If you still think there is some error, let me know. I might be wrong. But based on the experiments and debugging which I did, copy() was working.

**Comment by ayu-22** [comment]
@qubvel I have made the necessary changes. I have also tried to train the network, and the model was training well like, there was a decrease in loss and the decrease becomes less after every epoch, etc.

But, I am not able to find a way to update the doc. It would be helpful if you guide me on that. I had updated the README already

**Comment by qubvel** [comment]
> Currently, I am doing the comparison with respect to Unet. I will let you know, once I am done with experiments

@ayu-22 did you have some experiments results? Graphs with losses and metics with comparison to Unet?

**Comment by ayu-22** [comment]
@qubvel I have trained the model to few epochs. It was training well i.e. the loss was decreasing and validation accuracy/iou was increasing.
I do not have the comparison with unet and because of GPU constraints, it may take some time. If you want I start the training of unet and give the comparison results.

**Review by seefun — APPROVED** [review]

**Comment by Rusab** [comment]
Sad that we still don't have Unet3+ in the library
